### PR TITLE
Add Dockerfile (for Umbrel)

### PR DESCRIPTION
**Pull request by nickfarrow**
Resolves https://github.com/chaincase-app/nolooking/issues/4

### CODE REVIEW

**Comment by DanGould** [comment]
Umbrel requires images to be built from debian:buster-slim. Is alpine a subset of that image?

**Comment by nickfarrow** [comment]
Ill try switch it over, hopefully builds!

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+RUN echo "Called build!"
+
+# Run Loin from a final debian container
+FROM debian:buster-slim
+COPY --chown=1000:1000 . .
+# USER 1000
+# Copy just the binary from our build stage
+COPY --from=chosen_builder /usr/local/cargo/bin/loin /usr/local/bin/loin
+COPY run_loin /usr/local/bin/run_loin
+COPY static/ ./static/
```
Comment (DanGould) [inline Dockerfile:57]:
```suggestion
COPY static/ /usr/share/loin/static/
```

File: run_loin [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+#     echo "lnd_address=\"https://lightning_lnd_1:$LND_GRPC_PORT\"" >> $CONF
+#     echo "lnd_cert_path=\"$TLS_FILE\"" >> $CONF
+#     echo "lnd_macaroon_path=\"$MACAROON_FILE\"" >> $CONF
+#     echo "endpoint=\"https://$APP_HIDDEN_SERVICE\"" >> $CONF
+# fi
+# cat $CONF
+#loin --conf $CONF
+
+echo "Running loin"
+loin --bind-port=4444 --lnd-address=https://lightning_lnd_1:$LND_GRPC_PORT --lnd-cert-path=$TLS_FILE --lnd-macaroon-path=$MACAROON_FILE --endpoint=https://$APP_HIDDEN_SERVICE
```
Comment (DanGould) [inline run_loin:18]: Could this change? Would this ip still work over https? I think we want as few hard-coded variables as possible

```suggestion
loin --bind-port=4444 --lnd-address=https://$APP_LIGHTNING_IP:$LND_GRPC_PORT --lnd-cert-path=$TLS_FILE --lnd-macaroon-path=$MACAROON_FILE --endpoint=https://$APP_HIDDEN_SERVICE
```

**Comment by DanGould** [comment]
Why is this a draft? Aren't [`d51af09`](https://github.com/chaincase-app/nolooking/pull/22/commits/d51af099a707ec03c519c1416c182d419bd2ece5) and [`44a6d80`](https://github.com/chaincase-app/nolooking/pull/22/commits/44a6d806c14b1629fa18e5d7f69eb50c297d9998) in the release?

**Comment by nickfarrow** [comment]
Yep, built upon those and also https://github.com/chaincase-app/nolooking/pull/49

**Comment by DanGould** [comment]
The prereq PRs are both merged. looks like we're ready to roll a 0.1.1 when this is in!

**Comment by 0xBEEFCAF3** [comment]
> @0xBEEFCAF3 could you check this on Umbrel?

Yes I can do this today

**Comment by 0xBEEFCAF3** [comment]
Build failed
Std err logs: 
```
error: failed to run custom build command for `ring v0.16.20`

Caused by:
  process didn't exit successfully: `/usr/src/nolooking/target/release/build/ring-05beccad6314eebf/build-script-build` (exit status: 101)
  --- stdout
  OPT_LEVEL = Some("3")
  TARGET = Some("x86_64-unknown-linux-musl")
  HOST = Some("aarch64-unknown-linux-gnu")
  cargo:rerun-if-env-changed=CC_x86_64-unknown-linux-musl
  CC_x86_64-unknown-linux-musl = None
  cargo:rerun-if-env-changed=CC_x86_64_unknown_linux_musl
  CC_x86_64_unknown_linux_musl = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  RUSTC_LINKER = None
  cargo:rerun-if-env-changed=CROSS_COMPILE
  CROSS_COMPILE = None
  cargo:rerun-if-env-changed=CFLAGS_x86_64-unknown-linux-musl
  CFLAGS_x86_64-unknown-linux-musl = None
  cargo:rerun-if-env-changed=CFLAGS_x86_64_unknown_linux_musl
  CFLAGS_x86_64_unknown_linux_musl = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("false")
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,llvm14-builtins-abi,sse,sse2")

  --- stderr
  running "musl-gcc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-I" "include" "-Wall" "-Wextra" "-pedantic" "-pedantic-errors" "-Wall" "-Wextra" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wenum-compare" "-Wfloat-equal" "-Wformat=2" "-Winline" "-Winvalid-pch" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wundef" "-Wuninitia
lized" "-Wwrite-strings" "-fno-strict-aliasing" "-fvisibility=hidden" "-fstack-protector" "-g3" "-U_FORTIFY_SOURCE" "-DNDEBUG" "-c" "-o/usr/src/nolooking/target/x86_64-unknown-linux-musl/release/build/ring-5921cfd37d58aa5a/out/aesni-x86_64-elf.o" "/usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.20/pregenerated/aesni-x86_64-elf.S"
  cc1: error: unrecognized command line option '-m64'
  thread 'main' panicked at 'execution failed', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.20/build.rs:656:9
  stack backtrace:
     0: std::panicking::begin_panic
     1: build_script_build::run_command
     2: build_script_build::compile
     3: build_script_build::build_library::{{closure}}
     4: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
     5: core::option::Option<T>::map
     6: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
     7: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
     8: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
     9: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
    10: core::iter::traits::iterator::Iterator::collect
    11: build_script_build::build_library
    12: build_script_build::build_c_code::{{closure}}
    13: <core::slice::iter::Iter<T> as core::iter::traits::iterator::Iterator>::for_each
    14: build_script_build::build_c_code
    15: build_script_build::ring_build_rs_main
    16: build_script_build::main
    17: core::ops::function::FnOnce::call_once
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
warning: build failed, waiting for other jobs to finish...
error: failed to compile `nolooking v0.1.0 (/usr/src/nolooking)`, intermediate artifacts can be found at `/usr/src/nolooking/target`
The command '/bin/sh -c cargo install --target x86_64-unknown-linux-musl --path .' returned a non-zero code: 101
```

**Comment by DanGould** [comment]
@0xBEEFCAF3 that looks like a cargo error to me. I was able to build on my machine, but I just can't deploy it on an umbrel.

@nickfarrow what do you think is the best test for "passing" this script in the future? I'd like to build a container for armv7 in this Dockerfile in the future, too

**Comment by nickfarrow** [comment]
What machine are you trying to build on? Mine is building from x86 with commands:

```
sudo docker run --privileged --rm tonistiigi/binfmt --install arm64,riscv64,arm
```

then
```
sudo docker buildx build --platform linux/arm64,linux/amd64 --tag satsale/nolooking:IMGNUMBER --output type=registry .
```
Maybe those will help?

**Comment by DanGould** [comment]
@nickfarrow I was doing it wrong. I just ran

```
docker buildx create --name umbrel
docker buildx use umbrel
docker buildx inspect --bootstrap
```

to get the builder. before I checked the logs and had only built the x86_64 version. I will put what's building now on dockerhub and then @0xBEEFCAF3  can run it from there as a smoke test

**Comment by DanGould** [comment]
```console
 => [linux/arm64 branch-version-arm64 3/4] RUN rustup target add aarch64-unknown-linux-musl                                                                          34.8s
 => ERROR [linux/arm64 branch-version-arm64 4/4] RUN cargo install --target aarch64-unknown-linux-musl --path .                                                     188.1s
------
 > [linux/arm64 branch-version-arm64 4/4] RUN cargo install --target aarch64-unknown-linux-musl --path .:
#0 4.471   Installing nolooking v0.1.0 (/usr/src/nolooking)
#0 19.38     Updating crates.io index
#0 187.7 Killed
------
Dockerfile:43
--------------------
  41 |     ENV AR_aarch64_unknown_linux_musl=llvm-ar
  42 |     ENV CARGO_TARGET_AARCH64_UNKNOWN_LINUX_MUSL_RUSTFLAGS="-Clink-self-contained=yes -Clinker=rust-lld"
  43 | >>> RUN cargo install --target aarch64-unknown-linux-musl --path .
  44 |     
  45 |     # We build for either x86_64 or ARM from above options using the docker $TARGETARCH
--------------------
ERROR: failed to solve: process "/bin/sh -c cargo install --target aarch64-unknown-linux-musl --path ." did not complete successfully: exit code: 137
```

**Comment by nickfarrow** [comment]
Are you sure you ran the `--install arm64` command? Either that or out of ram

**Comment by DanGould** [comment]
yes I `--install arm64`

also 16gb ram machine

**Comment by DanGould** [comment]
still runs out of ram 🤔 why does it need so much
image

**Comment by nickfarrow** [comment]
I was thinking that this arm build just requires a lot of memory to produce (all the statically linked dependencies), but maybe there is something else going wrong?

Using shared local dependencies outside of docker may help, but i haven't tried yet
https://stackoverflow.com/questions/54952867/cache-cargo-dependencies-in-a-docker-volume

**Comment by DanGould** [comment]
Finally got it built and [on the chaincase dockerhub](https://hub.docker.com/layers/chaincase/nolooking/0.1.1-alpha.rc3/images/sha256-5328d5f6a41ee0bc5beef5a035dc14c524206f44818bf098d77e7cd552d79630?context=repo) 

Apparently there is a big memory leak in libgit2, a c library cargo depends on. Rust can still leak memory if it depends on c code 🤣 By passing `--config "net.git-fetch-with-cli=true"` to `cargo install` I was able to compile on aarch64. It took almost an hour. Big thanks to @asottile-sentry for finding the bug and making a video on it.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+FROM builder AS branch-version-arm64
+RUN echo "Preparing to cargo build for arm (${TARGETARCH})"
+# Install the required dependencies to build for `musl` static linking for arm.
+RUN apt-get update && apt-get install musl-tools clang llvm -y
+# Add our arm target to rust, some build variables, then compile and install
+RUN rustup target add aarch64-unknown-linux-musl
+ENV CC_aarch64_unknown_linux_musl=clang
+ENV AR_aarch64_unknown_linux_musl=llvm-ar
+ENV CARGO_TARGET_AARCH64_UNKNOWN_LINUX_MUSL_RUSTFLAGS="-Clink-self-contained=yes -Clinker=rust-lld"
+RUN cargo install --target aarch64-unknown-linux-musl --path .
```
Comment (DanGould) [inline Dockerfile:43]:
```suggestion
RUN cargo --config "net.git-fetch-with-cli=true" install --target x86_64-unknown-linux-musl --path .

# ARM
FROM builder AS branch-version-arm64
RUN echo "Preparing to cargo build for arm (${TARGETARCH})"
# Install the required dependencies to build for `musl` static linking for arm.
RUN apt-get update && apt-get install musl-tools clang llvm -y
# Add our arm target to rust, some build variables, then compile and install
RUN rustup target add aarch64-unknown-linux-musl
ENV CC_aarch64_unknown_linux_musl=clang
ENV AR_aarch64_unknown_linux_musl=llvm-ar
ENV CARGO_TARGET_AARCH64_UNKNOWN_LINUX_MUSL_RUSTFLAGS="-Clink-self-contained=yes -Clinker=rust-lld"
RUN cargo --config "net.git-fetch-with-cli=true" install --target aarch64-unknown-linux-musl --path .
```

**Review by DanGould — CHANGES_REQUESTED** [review]

**Pull request merged**